lib/bap_disasm/bap_disasm_rec.ml

```diff
@@ -241,34 +202,54 @@ let next dis s =
     | [] -> Dis.stop dis s
     | r :: roots when not(Memory.contains s.base r) ->
       loop {s with roots}
-    | r :: roots when Span.mem s.visited r ->
+    | r :: roots when Visited.is_insn s.visited r ->
       loop {s with roots}
     | addr :: roots ->
-      let mem = match Span.upper_bound s.visited addr with
-        | None -> Memory.view ~from:addr s.base
-        | Some r1 -> Memory.range s.base addr r1  in
-      let mem =
-        Result.map_error mem ~f:(fun err -> Error.tag err "next_root") in
-      mem >>= fun mem -> Dis.jump dis mem {s with roots; addr} in
-  let visited = Span.add s.visited (s.addr, Dis.addr dis) in
-  loop {s with visited}
+      let mem = Memory.view ~from:addr s.base in
+      let mem = Result.map_error mem ~f:(fun err -> Error.tag err "next_root") in
+      mem >>= fun mem ->
+      Dis.jump dis mem {s with roots; addr} in
+  loop s
 
 let stop_on = [`Valid]
 
+(* find intersections, i.e. instructions which share bytes, but have different address,
+   and update destinations for those of them, which also share max address to
+   split a block in such place *)
+let update_intersections disasm brancher s =
+  let next dis s = match s.roots with
+    | [] -> Dis.stop dis s
+    | addr :: roots ->
+      Memory.view ~from:addr s.base >>= fun mem ->
+      Dis.jump dis mem {s with roots; addr} in
+  let roots,intersected = find_intersections s in
+  let s = {s with intersected; roots=Set.to_list roots} in
```

Comment:
why `Set.to_list`? Is there any particular reason to change the representation? If not, then please keep it as Set.t

Reply:
Set is not sufficient, because we need to know both first and last addresses of an instruction to figure out if it's a block terminator or not. `Bap.Std.Table` doesn't fit for this particular task because we have intersected memory regions. We can also use `Bap.Std.Memmap` in most of the cases, but it won't be so convenient as just the `addr->addr` mapping since we use addresses everywhere as a key for tables. And sometimes it will be even hard to use `mem` as a key when we need to infer the next key from a given one, which is easy to do in case of address by incrementing it, and not so easy for memory chunk.